### DISCUSSION TITLE
Natively handle inverted mouse tracker

### DISCUSSION THREAD

**Discussion by geoff-harper** [openseadragon/openseadragon#2224] [Ideas]
I've been using OSD for a few years now and I'm one of the seemingly many people that want right mouse button to pan. I know there are examples floating around of strapping an extra mousetracker onto the viewer and doing it in userland, but it'd be really great if we could use the internal mouse trackers to handle this. I also have experience with Three.js (3D webgl lib) and they have some controls out of the box with configurable [mouse](https://threejs.org/docs/index.html#examples/en/controls/OrbitControls.mouseButtons) and [gesture](https://threejs.org/docs/index.html#examples/en/controls/OrbitControls.touches) controls. 

I don't think we really need the same level as abstraction even, just a parameter in the Viewer constructor where we could pass in something like `primaryButton: 'LEFT' | 'RIGHT' | 'MIDDLE'` (if middle is not supported for primary just omit that one). Then all further calls that emit the primary/non-primary events would check equality against that property instead of explicitly `0 | 1`. If there's interest in this I can take a stab at it but my bitwise operator knowledge is very rusty (re: `updatePointerDown`)

**Comment by iangilman** [comment]
Seems like a fine feature to add. Would we really want to re-map the primary button for everything, or is it more that we just want panning to be done via the right mouse? The other mouse button gesture I can think of is click (and/or double click) to zoom. Would we want those to be based on the right mouse button? My intuition is we wouldn't necessarily always want them both re-mapped. Maybe it would be good to add something to the gesture settings like `panWithSecondaryButton`. 

In terms of terminology, I think we've gone with "secondary", rather than "right", partially because I think for pens there can be a secondary button that's not necessarily on the right. Also, of course, left handed people can re-map their buttons so the secondary button is on the left. @msalsbery is that correct? 

As for supporting the middle button, I guess I would say let's only do it if we anticipate actually wanting it. That said, the gesture settings should be something like `panButton` that can be one of three options in the future if we decide (so at least we aren't cutting off that option for later). 

@geoff-harper If you're up for taking a whack at this, please do! Let me know if you need anything along the way...

**Comment by geoff-harper** [comment]
I think if pan remapping is added in, requests to remap click zoom will inevitably follow. In my mind it's either remap the primary or a full blown keybind matrix. A keybind matrix seems ambitious to start so maybe just primary remap? I think if users really want only the pan flipped then that's still an option in userland. Although I'll defer to your judgement on it, personally I have both click to zoom functions disabled since we require clicking on another canvas overlayed on top

[W3C has main, secondary, and auxiliary](https://w3c.github.io/uievents/#dom-mouseevent-button) so I'm keen to follow that verbatim.

**Comment by iangilman** [comment]
My concern with remapping everything is I don't want to paint ourselves into a corner with the API. If we go with `primaryButton` now, and it remaps both click to zoom and drag to pan, if in the future we want to differentiate between click and drag, what do we do with the `primaryButton` API? If, instead, we start with `panButton` now, and later we decide we want to be able to remap click to zoom, we can add `clickButton` and nothing needs to change with `panButton`.

So yeah, my recommendation is we just focus on remapping pan for now, because of that reason, and because it's the only thing you need at this time. 

We are also using main, secondary, and auxiliary internally and in some of the docs, so I'm happy to continue that.

**Comment by msalsbery** [comment]
> Seems like a fine feature to add. Would we really want to re-map the primary button for everything, or is it more that we just want panning to be done via the right mouse? The other mouse button gesture I can think of is click (and/or double click) to zoom. Would we want those to be based on the right mouse button? My intuition is we wouldn't necessarily always want them both re-mapped. Maybe it would be good to add something to the gesture settings like `panWithSecondaryButton`.
> 
> In terms of terminology, I think we've gone with "secondary", rather than "right", partially because I think for pens there can be a secondary button that's not necessarily on the right. Also, of course, left handed people can re-map their buttons so the secondary button is on the left. @msalsbery is that correct?
> 
> As for supporting the middle button, I guess I would say let's only do it if we anticipate actually wanting it. That said, the gesture settings should be something like `panButton` that can be one of three options in the future if we decide (so at least we aren't cutting off that option for later).
> 
> @geoff-harper If you're up for taking a whack at this, please do! Let me know if you need anything along the way...

@iangilman Yes that's correct!

I'm taking a look at this. To make the behavior best on the viewer, MouseTracker should be involved in this as well, mostly for pointer capture which is desired in pretty much all pointer-down/gesture-from-pointer-move/pointer-up situations

**Reply by iangilman** [reply]
@msalsbery Great! Thank you for taking a look at this :)
